ocaml/xapi/helpers.ml

```diff
@@ -1234,3 +1234,108 @@ let host_supports_hvm ~__context host =
 let env_with_path env_vars =
   Array.of_list @@
   List.map (fun (k,v) -> Printf.sprintf "%s=%s" k v) (("PATH", String.concat ":" Forkhelpers.default_path)::env_vars)
+
+(** wrapper around the stunnel@xapi systemd service.
+  * there exist scripts (e.g. xe-toolstack-restart) which also manipulate
+    the stunnel daemon but they do this directly (not via ocaml). *)
+module Stunnel : sig
+  (** restart stunnel, possibly changing the config file *)
+  val restart : __context:Context.t -> accept:string -> unit
+end = struct
+
+  let cert = !Xapi_globs.server_cert_path
+
+  (** protect ourselves from concurrent writes to files *)
+  module File : sig
+    (** create or update stunnel config file *)
+    val update_config : accept:string -> unit
+    (** calls `generate_ssl_cert` bash script
+      * TODO: don't generate certificates in bash *)
+    val gen_pem_if_not_exist : __context:Context.t -> unit
+  end = struct
+    let xapi_stunnel_conf = "/etc/stunnel/xapi.conf"
+    let m = Mutex.create ()
+    let current_accept = ref None
+
+    let rewrite_xapi_ssl_config_file ~accept =
+      let idle_timeout = match Sys.getenv_opt "STUNNEL_IDLE_TIMEOUT" with
+                         | None -> "; no idle timeout"
+                         | Some x -> Printf.sprintf "TIMEOUTidle = %s" x
+      in
+      let conf_contents = String.concat "\n" [
+          "; autogenerated by helpers.ml"
+        ; "fips = no"
+        ; "pid = /var/run/xapissl.pid"
+        ; "socket = r:TCP_NODELAY=1"
+        ; "socket = a:TCP_NODELAY=1"
+        ; "socket = l:TCP_NODELAY=1"
+        ; "socket = r:SO_KEEPALIVE=1"
+        ; "socket = a:SO_KEEPALIVE=1"
+        ; idle_timeout
+        ; "debug = authpriv.5"
+        ; ""
+        ; "[xapi]"
+        ; Printf.sprintf "accept = %s" accept
+        ; "connect = 80"
+        ; Printf.sprintf "cert = %s" cert
+        ; Printf.sprintf "ciphers = %s" Xcp_const.good_ciphersuites
+        ; "curve = secp384r1"
+        ; "TIMEOUTclose = 1"
+        ; "options = CIPHER_SERVER_PREFERENCE"
+        ; "sslVersion = TLSv1.2"
+        ]
+      in
+      let len = String.length conf_contents in
+      Stdext.Unixext.atomic_write_to_file xapi_stunnel_conf 0o0600
+        (fun fd -> Unix.single_write_substring fd conf_contents 0 len |> ignore)
+
+    let update_config ~accept =
+      Mutex.execute m (fun () ->
+        match !current_accept with
+        | Some current_accept when current_accept = accept -> ()
+        | None | Some _ -> begin
+          current_accept := Some accept;
+          rewrite_xapi_ssl_config_file ~accept
+        end
+      )
+
+    let gen_pem_if_not_exist ~__context =
```

Comment:
I'd be interested in triggering this function, when there's a certificate present but it's expired.

One option to allow for this would be to provide the file check as a parameter to move it outside if the function (restart would provide the check)

I'm open to other ideas as this isn't super clean, but the handling of the mutex makes it tricky.

Reply:
it would probably make sense if you became the owner of this function, and this stunnel helper becomes a client (or we figure out a way to decouple restarting stunnel and writing the stunnel config completely)

that way we you could handle concurrent writes yourself without the confusion of using this mutex